First steps of the Solang Solidity Server

Hyperion101010 opened the pull request.

seanyoung commented on .vscode/extensions.json:
```diff
@@ -0,0 +1,7 @@
+	"recommendations": [
+		"dbaeumer.vscode-eslint"
+	]
+}
```
This file is not valid json. Also, it lists a set of extensions to be recommended with ours. I do not think we can recommend anything so please delete the file.

seanyoung commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,24 @@
+    - name: Run tests
+      uses: GabrielBB/xvfb-action@v1.2
+      with:
+        run: npm test
```
What happened to the rust.yaml?
For eval 1 we have send message to rust server so rust must be compiled.

seanyoung commented on .vscode/settings.json:
```diff
@@ -0,0 +1,11 @@
+    },
+    // Turn off tsc task auto detection since we have the necessary tasks as npm scripts
+    "typescript.tsc.autoDetect": "off"
+}
```
This file not valid json. Unless we have some vscode settings for the project, remove

seanyoung commented on solidity.configuration.json:
```diff
@@ -0,0 +1,25 @@
+		["[", "]"],
+		["(", ")"]
+	]
+}
```
These files are from https://github.com/juanfranblanco/vscode-solidity/ (I think!)

Proper attribution should happen (saying where it came from (url), what the license is, and who the copyright holder is).

This can be in the file itself or a separate file in snippets/

seanyoung commented on src/client/extension.ts:
```diff
@@ -0,0 +1,161 @@
+			transport: TransportKind.ipc,
+		}
+	};
+	*/
```
Please remove commented out code.

seanyoung commented on src/dummy.js:
```diff
@@ -0,0 +1,19 @@
+const chld_proc = require('child_process');
+
+var som = chld_proc.spawn('cargo run',['--example', 'goto_def'], {cwd:'/home/hyperion/intern/hyperledger/sls/practice/lsp-server', shell:true});
+console.log(som);
```
Is this file used anywhere? Please delete. It refers to paths on your machine.

seanyoung commented on src/test/suite/helper.ts:
```diff
@@ -0,0 +1,61 @@
+	try {
+		doc = await vscode.workspace.openTextDocument(docUri);
+		editor = await vscode.window.showTextDocument(doc);
+		await sleep(10000); // Wait for server activation
```
We need a better solution rather than a random 10 second wait.

seanyoung commented on syntaxes/solidity.json:
```diff
@@ -0,0 +1,540 @@
+    },
+    "scopeName": "source.solidity",
+    "uuid": "123"
+}
```
This file needs proper attribution. We also need to know what version was copied into our repo, so we can update it. Please refer to the exact commit of the source repo and include copyright and license information.

seanyoung reviewed (changes requested):
Looks great, some minor review comments

Commit:
First steps of the Solang Solidity Server.

 Implements hover provider, server for LSP, syntax highlighting,
 test for apply edit.
 This server is able to execute commands from client, open/close/save
 files.
 Uses Solidity grammer from existing Solidity references(references in
 the README)
 To get started with the extension follow README.

Signed-off-by: Shivam Balikondwar <shivambalikondwar@icloud.com>
